allennlp/nn/beam_search.py

```diff
@@ -222,6 +223,14 @@ def _search(
                 start_predicted_classes,
             ) = start_class_log_probabilities.topk(self.beam_size)
 
+        # Some samplers add perturb the log probabilities with random noise, such
+        # as the GumbelMaxSampler, so we need to recover and keep track of the
+        # true log probabilities here.
+        if self.sampler is not None:
+            last_true_probabilities = torch.gather(
+                start_class_log_probabilities, 1, start_predicted_classes
+            )
```

Comment:
I think it's better to define this variable outside of this branch. I'm surprised Python allows this kind of thing, but then again... it's Python so I guess anything goes 😝

```suggestion
        last_true_probabilities: Optional[torch.Tensor] = None
        if self.sampler is not None:
            last_true_probabilities = torch.gather(
                start_class_log_probabilities, 1, start_predicted_classes
            )
```

Reply:
Also, should this be `last_true_log_probabilities`, not `last_true_probabilities`?

Reply:
Yup I'll change that variable name!